components/layout/construct.rs

```diff
@@ -1923,3 +1849,189 @@ impl ComputedValueUtils for ServoComputedValues {
            border.border_left_width != Au(0)
     }
 }
+
+/// Maintains a stack of anonymous boxes needed to ensure that the flow tree is *legal*. The tree
+/// is legal if it follows the rules in CSS 2.1 § 17.2.1.
+///
+/// As an example, the legalizer makes sure that table row flows contain only table cells. If the
+/// flow constructor attempts to place, say, a block flow directly underneath the table row, the
+/// legalizer generates an anonymous table cell in between to hold the block.
+///
+/// Generally, the flow constructor should use `Legalizer::add_child()` instead of calling
+/// `Flow::add_new_child()` directly. This ensures that the flow tree remains legal at all times
+/// and centralizes the anonymous flow generation logic in one place.
+struct Legalizer {
+    stack: Vec<FlowRef>,
+}
+
+impl Legalizer {
+    fn new() -> Legalizer {
+        Legalizer {
+            stack: vec![],
+        }
+    }
+
+    fn add_child(&mut self, stylist: &Stylist, parent: &mut FlowRef, mut child: FlowRef) {
+        while !self.stack.is_empty() {
+            if self.try_to_add_child(stylist, parent, &mut child) {
+                return
+            }
+            self.flush_top_of_stack(parent)
+        }
+
+        while !self.try_to_add_child(stylist, parent, &mut child) {
+            self.push_next_anonymous_flow(stylist, parent)
+        }
+    }
+
+    /// Flushes all flows we've been gathering up.
+    fn finish(mut self, parent: &mut FlowRef) {
+        while !self.stack.is_empty() {
+            self.flush_top_of_stack(parent)
+        }
+    }
+
+    /// Returns true if the child was successfully added and false otherwise.
+    fn try_to_add_child(&mut self, stylist: &Stylist, parent: &mut FlowRef, child: &mut FlowRef)
+                        -> bool {
+        let mut parent = self.stack.last_mut().unwrap_or(parent);
+        let (parent_class, child_class) = (parent.class(), child.class());
+        match (parent_class, child_class) {
+            (FlowClass::TableWrapper, FlowClass::Table) |
+            (FlowClass::Table, FlowClass::TableColGroup) |
+            (FlowClass::Table, FlowClass::TableRowGroup) |
+            (FlowClass::Table, FlowClass::TableRow) |
+            (FlowClass::Table, FlowClass::TableCaption) |
+            (FlowClass::TableRowGroup, FlowClass::TableRow) |
+            (FlowClass::TableRow, FlowClass::TableCell) => {
+                parent.add_new_child((*child).clone());
+                true
+            }
+
+            (FlowClass::Flex, FlowClass::Block) |
+            (FlowClass::Flex, FlowClass::Flex) => {
```

Comment:
Oh, I mean IMO the wrapper block should only be needed for `FlowClass::Inline` since other children are guaranteed to be block-like.